include/tvm/relay/base.h

```diff
@@ -0,0 +1,240 @@
+  void VisitAttrs(AttrVisitor* v) final {
+    v->Visit("source", &source);
+    v->Visit("lineno", &lineno);
+    v->Visit("col_offset", &col_offset);
+  }
+
+  TVM_DLL static Span make(SourceName source, int lineno, int col_offset);
+
+  static constexpr const char* _type_key = "relay.Span";
+  TVM_DECLARE_NODE_TYPE_INFO(SpanNode, Node);
+};
+
+RELAY_DEFINE_NODE_REF(Span, SpanNode, NodeRef);
+
+/*!
+ * \brief This is the base node container of all relay structures.
+ */
+class RelayNode : public Node {
+ public:
+  /*! \brief The location of the program in a SourceFragment can be null,
+   * check with span.defined() */
+  mutable Span span;
+
+  static constexpr const char* _type_key = "relay.Node";
+  TVM_DECLARE_BASE_NODE_INFO(RelayNode, Node);
+};
+
+/*!
+ * \brief Get a reference type from a Node ptr type
+ *
+ *  It is always important to get a reference type
+ *  if we want to return a value as reference or keep
+ *  the node alive beyond the scope of the function.
+ *
+ * \param ptr The node pointer
+ * \tparam RefType The reference type
+ * \tparam NodeType The node type
+ * \return The corresponding RefType
+ */
+template <typename RefType, typename NodeType>
+RefType GetRef(const NodeType* ptr) {
+  static_assert(std::is_same<typename RefType::ContainerType, NodeType>::value,
+                "Can only cast to the ref of same container type");
+  return RefType(const_cast<NodeType*>(ptr)->shared_from_this());
+}
+
+// TODO(@tqchen, @jroesch): can we move these semantics to HalideIR
+template <typename T>
+inline const T* As(const NodeRef& node) {
+  const Node* ptr = static_cast<const Node*>(node.get());
+  if (ptr && (ptr->is_type<T>() || ptr->derived_from<T>())) {
+    return static_cast<const T*>(ptr);
+  }
+  return nullptr;
+}
+
+template <typename T, typename U>
+std::vector<T> Downcast(std::vector<U> array) {
+  std::vector<T> out;
+  for (const U& elem : array) {
+    const typename T::ContainerType* node =
+        elem.template as<typename T::ContainerType>();
+    CHECK(node) << "Downcast failed" << std::endl;
+    out.push_back(GetRef<T>(node));
+  }
+  return out;
+}
+
+template <typename T, typename U>
+Array<T> Downcast(Array<U> array) {
+  Array<T> out;
+  for (const U& elem : array) {
+    const typename T::ContainerType* node =
+        elem.template as<typename T::ContainerType>();
+    CHECK(node) << "Downcast failed" << std::endl;
+    out.push_back(GetRef<T>(node));
+  }
+  return out;
+}
+
+template <typename SubRef, typename BaseRef>
+SubRef Downcast(BaseRef ref) {
+  const typename SubRef::ContainerType* node =
+        ref.template as<typename SubRef::ContainerType>();
+    CHECK(node) << "Downcast failed" << std::endl;
+    return GetRef<SubRef>(node);
+}
+
+/*!
+ * \brief Get PackedFunction from global registry and
+ *  report error if it does not exist
+ * \param name The name of the function.
+ * \return The created PackedFunc.
+ */
+inline const PackedFunc& GetPackedFunc(const std::string& name) {
```

Comment:
remove this for now? Seems we can simply inline it at caller to avoid one level of indriection